llama_index/llm_predictor/huggingface.py

```diff
@@ -0,0 +1,229 @@
+        if "cuda" == self._device_map or (
+            "auto" == self._device_map and torch.cuda.is_available()
+        ):
+            inputs = inputs.to("cuda")
+
+        streamer = TextIteratorStreamer(
+            self.tokenizer,
+            skip_prompt=True,
+            decode_kwargs={"skip_special_tokens": True},
+        )
+        generation_kwargs = dict(
+            inputs,
+            streamer=streamer,
+            max_new_tokens=self._max_new_tokens,
+            temperature=self._temperature,
+            do_sample=self._do_sample,
+            stopping_criteria=self._stopping_criteria,
+        )
+
+        # generate in background thread
+        # NOTE/TODO: token counting doesn't work with streaming
+        thread = Thread(target=self.model.generate, kwargs=generation_kwargs)
+        thread.start()
+
+        # create generator based off of streamer
+        def response() -> Generator:
+            for x in streamer:
+                yield x
+
+        return response(), formatted_prompt
+
+    @property
+    def total_tokens_used(self) -> int:
+        """Get the total tokens used so far."""
+        return self._total_tokens_used
+
+    @property
+    def last_token_usage(self) -> int:
+        """Get the last token usage."""
+        return self._last_token_usage or 0
+
+    @last_token_usage.setter
+    def last_token_usage(self, value: int) -> None:
+        """Set the last token usage."""
+        self._last_token_usage = value
+
+    def predict(self, prompt: Prompt, **prompt_args: Any) -> Tuple[str, str]:
+        """Predict the answer to a query.
+
+        Args:
+            prompt (Prompt): Prompt to use for prediction.
+
+        Returns:
+            Tuple[str, str]: Tuple of the predicted answer and the formatted prompt.
+
+        """
+        import torch
+
+        formatted_prompt = prompt.format(**prompt_args)
+        full_prompt = self._query_wrapper_prompt.format(query_str=formatted_prompt)
+        if self._system_prompt:
+            full_prompt = f"{self._system_prompt} {full_prompt}"
+
+        inputs = self.tokenizer(full_prompt, return_tensors="pt")
+        if "cuda" == self._device_map or (
+            "auto" == self._device_map and torch.cuda.is_available()
+        ):
+            inputs = inputs.to("cuda")
+
+        tokens = self.model.generate(
+            **inputs,
+            max_new_tokens=self._max_new_tokens,
+            temperature=self._temperature,
+            do_sample=self._do_sample,
+            stopping_criteria=self._stopping_criteria,
+        )
+        completion_tokens = tokens[0][inputs["input_ids"].size(1) :]
+        self._total_tokens_used += len(completion_tokens) + inputs["input_ids"].size(1)
+        completion = self.tokenizer.decode(completion_tokens, skip_special_tokens=True)
+        return completion, formatted_prompt
+
+    async def apredict(self, prompt: Prompt, **prompt_args: Any) -> Tuple[str, str]:
```

Comment:
curious: huggingface models don't expose async endpoints in general?

Reply:
Not that i saw. All examples put the normal code into an async function 🤔

Reply:
Good to know